### ISSUE TITLE
Regression with bound TypeVar in constructors

### ISSUE THREAD

**Issue by vberlier** [microsoft/pyright#2043]
This is a tricky one. With the latest Pyright release the following code no longer works:

```python
K = TypeVar("K")
V = TypeVar("V")

@dataclass
class Registry(Generic[K, V]):
    """Some type of generic registry."""

    key: K
    value: Dict[str, V] = field(default_factory=dict)

class Base:
    """An arbitrary base class for things to be stored in a specialized registry."""

BaseType = TypeVar("BaseType", bound=Base)

class BaseTypeRegistry(Registry[Type[BaseType], BaseType]):
    """A specialized registry for storing `Base` objects."""

class Thing1(Base):
    pass

class Thing2(Base):
    pass

reveal_type(BaseTypeRegistry(Thing1))  # BaseTypeRegistry[Type[Thing1]]

foo: BaseTypeRegistry[Thing1] = BaseTypeRegistry(Thing1)
# Expression of type "BaseTypeRegistry[Type[Thing1]]" cannot be assigned to declared type "BaseTypeRegistry[Thing1]"
#     TypeVar "BaseType@BaseTypeRegistry" is invariant
#       "Type[type]" is incompatible with "Type[Thing1]"
```

Until recently `reveal_type` would show that `BaseTypeRegistry(Thing1)` is of type `BaseTypeRegistry[Thing1]`, as expected. However with the latest version Pyright thinks that the type is `BaseTypeRegistry[Type[Thing1]]` instead.

**Comment by erictraut** [comment]
Thanks for reporting the bug, and apologies for the regression. I've incorporated a new test based on your same above so we can prevent similar regressions in the future.

**Comment by vberlier** [comment]
No worries. Thanks for the quick fix as always.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.155, which I just published. It will also be included in the next release of pylance.
